### ISSUE TITLE
Instancing new scene keeps variable from old scenes

### ISSUE THREAD

**Issue by tchernia** [godotengine/godot#33514]
**Godot version:**

3.1.1 - Win64

**OS/Device including version:**

Windows10

**Issue description:**

When I create a dictionary in the script of a scene, and then create several instances of the scene, the values added in the oldest instance are kept in the newest instances.

**Steps to reproduce:**

- Create both scenes with the corresponding script
- Click play

**Minimal reproduction project**

Here is a simple code with two scenes:

_Node scene:_
```
extends Node
var test_variable : Dictionary

# Called when the node enters the scene tree for the first time.
func _ready():
	test_variable[test_variable.size()] = "a"
	print("test variable: ")
	print(test_variable)
```

_Main scene:_
```
extends Node

var node_scene = preload("res://Node.tscn")

# Called when the node enters the scene tree for the first time.
func _ready():
	var a = node_scene.instance()
	add_child(a)
	
	var b = node_scene.instance()
	add_child(b)
	
	var c = node_scene.instance()
	add_child(c)
```

_Output:_
```
OpenGL ES 3.0 Renderer: AMD Radeon(TM) RX Vega 10 Graphics
test variable: 
{0:a}
test variable: 
{0:a, 1:a}
test variable: 
{0:a, 1:a, 2:a}
```

**Comment by YeldhamDev** [comment]
I think this was fixed in #32416. Could you test this with the latest beta?

**Comment by Xrayez** [comment]
Might be related to how arrays are[ shared between scenes](http://docs.godotengine.org/en/latest/getting_started/scripting/gdscript/gdscript_exports.html#exporting-arrays) according to docs:

> Exporting arrays works, but with an important caveat: while regular arrays are created local to every class instance, exported arrays are shared between all instances.

So it might be that dictionaries work the same way, but it isn't documented?

**Comment by Xrayez** [comment]
```gdscript
var test_variable : Dictionary
```
Make sure to initialize this as:
```gdscript
var test_variable := {}
# or
var test_variable := Dictionary()
```

**Comment by KoBeWi** [comment]
^
Yeah, this is a bug that was fixed in 3.2.

**Comment by tchernia** [comment]
Thank you very much for your very reactive answers, it is a pleasure to see how much the Godot community is willing to enhance the software and help :)

Indeed YeldhamDev and KoBeWi, i tested the same scene in Godot 3.2 and the issue was fixed there.

I am curious though, why was it happening? Is there a link with hidden "metadatas" as the issue you raised seem to point out? Or are there "class variables" (in that case scene variables)?

Xrayez thank you for the input. I think I will use that workaround since the 3.2 version is not yet stable. However I am not comfurtable since I don't understand what is going on here. In your proposition, we are not only declaring the dictionary but initializing it. Does that force the engine to build a new instance stored elsewhere in memory (I checked, and indeed if I try to edit one of the three dictionaries, the other two remains the same, thus proving these three dictionaries are stored in three different locations in memory)?

I will investigate the hypothesis concerning shared dictionaries. Indeed the issue is the same if I use an array instead of a dictionary. And the issue disappear when I am using an integer.
